From a5c9139861266d1f8d6a874d1fb087d9638e6076 Mon Sep 17 00:00:00 2001
From: Sandor Yu <Sandor.yu@nxp.com>
Date: Mon, 11 Dec 2017 17:10:48 +0800
Subject: [PATCH 3100/5242] MLK-17107-03: hdp: return unknow status if cable
 detect failed

commit  5043929a13f64456dc64344d80bdab289792ed5e from
https://source.codeaurora.org/external/imx/linux-imx.git

Return unknow status to connector if hdp cable detect function
failed to get cable status.

Signed-off-by: Sandor Yu <Sandor.yu@nxp.com>
(cherry picked from commit d16c633b1be1326b3583632c5279ea6b34ed3ec2)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-hdp.c |   13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.c b/drivers/gpu/drm/imx/hdp/imx-hdp.c
index b79332b..bf3fefc 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.c
@@ -648,7 +648,7 @@ static void imx_hdp_bridge_enable(struct drm_bridge *bridge)
 	struct imx_hdp *hdp = container_of(connector,
 						struct imx_hdp, connector);
 	int ret;
-	u8 hpd;
+	u8 hpd = 0xf;
 
 	ret = imx_hdp_call(hdp, get_hpd_state, &hdp->state, &hpd);
 	if (ret > 0)
@@ -657,9 +657,14 @@ static void imx_hdp_bridge_enable(struct drm_bridge *bridge)
 	if (hpd == 1)
 		/* Cable Connected */
 		return connector_status_connected;
-	else
-		/* Cable Disconnedted  */
+	else if (hpd == 0)
+		/* Cable Disconnedted */
 		return connector_status_disconnected;
+	else {
+		/* Cable status unknown */
+		DRM_INFO("Unknow cable status, hdp=%u\n", hpd);
+		return connector_status_unknown;
+	}
 }
 
 static int imx_hdp_connector_get_modes(struct drm_connector *connector)
@@ -984,7 +989,7 @@ static void hotplug_work_func(struct work_struct *work)
 		/* Cable Connected */
 		DRM_INFO("HDMI/DP Cable Plug In\n");
 		enable_irq(hdp->irq[HPD_IRQ_OUT]);
-	} else {
+	} else if (connector->status == connector_status_disconnected) {
 		/* Cable Disconnedted  */
 		DRM_INFO("HDMI/DP Cable Plug Out\n");
 		enable_irq(hdp->irq[HPD_IRQ_IN]);
-- 
1.7.9.5

